### ISSUE TITLE
PluginPattern支持nn.Module级别的匹配与替换

### ISSUE THREAD

**Issue by LeoLau94** [BD-Seed-HHW/xpu_graph#370]
### Describe the feature🚀

我们希望Plugin Pattern能够实现如下效果
```
                class TestModule(torch.nn.Module):
                    def __init__(self, *args, **kwargs):
                        super().__init__(*args, **kwargs)
                        self.w = torch.nn.Parameter(torch.rand(1024, 1024))
                        self.bias = torch.Tensor(1024)

                    def forward(self, x):
                        return (x @ self.w + self.bias,)

                class AnotherModule(torch.nn.Module):
                    def __init__(self, weight, *args, **kwargs):
                        super().__init__(*args, **kwargs)
                        self.w = torch.nn.Parameter(weight)

                    def forward(self, x):
                        return x + self.w

                weight = torch.rand(1024, 1024)
                register_plugin_pattern(
                    TestModule(), (torch.empty(1024, 1024),), AnotherModule(weight), target=Target.none
                )
                config = XpuGraphConfig(
                    is_training=False,
                    target=Target.none,
                    vendor_compiler_config={},
                    debug=True,
                    enable_cache=False,
                    freeze=True,
                )
                compiler = XpuGraph(config)

                input_tensor = torch.rand(1024, 1024)
                with torch.inference_mode():
                    compiled = torch.compile(TestModule(), fullgraph=True, dynamic=False, backend=compiler)
                    assert is_similar(compiled(input_tensor)[0], AnotherModule(weight)(input_tensor)) is True
```

### References

https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/graph_capture_wrappers.py#L1323

### Environment details

PyTorch: 2.5.1
torch_npu : 2.5.1
XpuGraph: 0.4.0

**Comment by LeoLau94** [comment]
我们注意到，默认情况下，dynamo会将本该是`get_attr`的Op，全部按照`placeholder`来处理

![Image](https://github.com/user-attachments/assets/e9bddbb4-8c4e-4f75-8cd0-31d29d0b75fe)

这是因为dynamo只有在使能[parameter freezing](https://github.com/pytorch/pytorch/blob/release/2.5/torch/_dynamo/utils.py#L3061)的情况下，才会按照`get_attr`来处理

在xpu_graph中，我们[提供了`freeze`和`is_training`来控制dynamo使能`parameter freezing`](https://github.com/BD-Seed-HHW/xpu_graph/blob/cd5dcfe64d08bce6adaa553795378901b37f4160/xpu_graph/compiler.py#L178)

这个问题的原因大抵如下：

在fx graph种，`get_attr`表示**从graph自身的Module中去取对应的`target`**，因此，如果使用`get_attr`，需要将对应的变量`浅拷贝`至GraphModule中。如下图所示

![Image](https://github.com/user-attachments/assets/82b34f44-8621-461c-be5b-208723f9380f)
实际上这大概等价于
```
graph_module._named_paramter['self__parameter_w`] = nn_module.w

```

“推理场景”的寓意一般来说是“无需梯度但不禁止用户替换对应的parameter/tensor”，这种情况下参数并不是真正的freezed。

目前存在“无需梯度”但“需要更新权重”的场景，如果dynamo还是使用`get_attr`的方式去读取对应的参数并且**用户没有使用原位操作来更新对应的参数**，那么Graph就没法得到最新的变量。

因此dynamo在处理这种情况，为了避免引入额外的用户使用限制（即必须使用原位操作来更新参数），不会使用`get_attr`，而是将其转换为`placeholder`，大抵等价于
```
graph_context.module = nn_module
graph_module.forward(graph_context.module.w)
```

这种限制还和异步并行相关，因为原位操作一般来说会引入同步操作。

**Comment by LeoLau94** [comment]
某种程度上，我们当然能够在Patter的fx graph里，将`get_attr`手动替换成`placeholder`节点，从而充分利用`placeholder`的通配性质，去匹配`paramter non-/freezing`两种场景。

可是，我们没法让`replacment`的`get_attr`全部换成placeholder，这是因为，即使replacement想要继承`pattern`中匹配的到的参数，在我们的issue给出的例子中，它一般也是由用户传入对应的参数并使用`__init__`构造好replacement module，再转交给我们。因此，我们会出现rewrite时，pattern和replacement没法一一对应的情况。


如果用户不能提前构造好replacement module，我们建议如上文那样，通过`funtional`方式去做match&rewrite，把module替换视为函数替换。

**Comment by LeoLau94** [comment]
```
weight = torch.Tensor()
module.weight = weight
graph_module.weight = module.weight
weight_1 = torch.Tensor()
module.weight = weight_1
assert graph_module.weight != module.weight
assert graph_module.weight == weight
# If we have & in python, then
weight = torch.Tensor()
module.weight = weight
graph_module.weight = &(module.weight)
weight_1 = torch.Tensor()
module.weight = weight_1
assert graph_module.weight == module.weight
assert graph_module.weight != weight
# That's why dynamo make get_attr as placeholder by default even it's in inference mode.
```

**Comment by LeoLau94** [comment]
目前的需求是，替换出来的pattern需要完成以下事情

1. 动态构造module对象，因为有一些信息来自于匹配的段落；
2. 调用module对象执行前推逻辑；

因此，我们的pattern匹配还需要抓取到构造module对象所需要的参数（作为placeholder表现在图里）
